src/block/block_read.c

```diff
@@ -221,6 +221,18 @@ __wt_block_read_off(WT_SESSION_IMPL *session, WT_BLOCK *block, WT_ITEM *buf, uin
     check_size = F_ISSET(&swap, WT_BLOCK_DATA_CKSUM) ? size : WT_BLOCK_COMPRESS_SKIP;
     if (swap.checksum == checksum) {
         blk->checksum = 0;
+
+        /*
+         * If chunk cache is configured we want to account for the race condition where the chunk
+         * cache could have stale content, and therefore a mismatched checksum. We do not want to
+         * fail here, free the stale content and read once so we can retry.
+         */
+        if (!__wt_checksum_match(buf->mem, check_size, checksum)) {
+            if (F_ISSET(&S2C(session)->chunkcache, WT_CHUNKCACHE_CONFIGURED)) {
+                WT_RET(__wt_chunkcache_free_external(session, block, objectid, offset, size));
+                WT_RET(__wt_read(session, block->fh, offset, size, buf->mem));
+            }
+        }
         if (__wt_checksum_match(buf->mem, check_size, checksum)) {
```

Comment:
Open to any input regarding the structuring of these if statements - struggling to find an implementation that is satisfying.

Reply:
I dislike that we're calling `__wt_checksum_match` twice even in the "success" case - this is potentially expensive.

Overall though, I don't mind the structure of this. So how about:
```
bool match = __wt_checksum_match(...);
if (!match) {
    ...
    match = __wt_checksum_match(...);
}
if (match) {
    ...
}
```

Reply:
good idea! done here [6d1b1da](https://github.com/wiredtiger/wiredtiger/pull/9838/commits/6d1b1da34e6d1213840588e9088bfc8f1c674881)